### ISSUE TITLE
Initial Roadmap

### ISSUE THREAD

**Issue by kindly** [kindly/flatterer-web#1]
### Use [libflatterer](https://github.com/kindly/libflatterer) to make an API that covers most aspects of flatterer:

* Using rust web framework to keep memory low and speed fast.
* Input sources can be: 
   * An external web URL.
   * JSON data in post request body.
   * Multipart form upload.
* File stored in tmp location and can be used across requests for limited time.
* Limits:
   * Size limit on upload/download should be fairly generous maybe 500MB
   * Upload/download time of files main concern will be larger than the processing time.
   * Files deleted after an hour.
   * Size of temporary space a contraint if used a lot. Possibly use LRU disk cache so time limit could be less than a hour if used a lot.
   * Memory limits **not** an issue as all can be run in single process with multiple threads working on all cores. 
   * Can easily run on heroku with sticky sessions, so that if multiple dynos are used then users will still see their data.
   * Probably condider some rate limiting, just so not abused.
* Output formats:
   * Zip File of whole output directory
   * Just XLSX File
   * Just main table as CSV
   * Just fields.csv file as JSON
   * Preview mode with first 10 lines of each output CSV file.

### Use API to build basic frontend:

* SPA probably in vue.
* Input sources can be: 
   * An external web URL.
   * Multipart form.
   * Text box in a form.
* If error in parsing the data or zero lines are created, suggest options that would mean file could be parsed.
* Initailly show preview of output CSV files.
* Give ways to edit/reorder the `fields.csv` in UI
* Buttons to download appropriate output if parsing went well.

**Comment by kindly** [comment]
@jpmckinney this could be of interest.

**Comment by jpmckinney** [comment]
Thank you for writing this up!

**Comment by kindly** [comment]
Initail prototype on heroku:    https://flatterer.herokuapp.com/

Has `jsonstream` guessing and the library allows top level objects, so most arbartary JSON should have ouput now without any config. For top level object it will stream any list of objects and store data in top level object in main table.  Meaning say for OCDS, it will handle packages whilst putting package data into the main table and putting releases/records in tables starting with `releases_`/`_records` if you do not specify a path.

**Comment by jpmckinney** [comment]
What is jsonstream guessing? Does that just mean it can handle a JSON stream (e.g. concatenated JSON) without a specific instruction?

**Comment by kindly** [comment]
yes it takes a sample from the top of the file and if the stream starts with object and that object ends and then there is another object starting directly after it assumes the input is a JSON stream.
